dk62_serhiienko/lab5_6_character_device/src/hivemod.c

```diff
@@ -0,0 +1,451 @@
+/* 
+* Based on @thodnev hivemod.c
+* The information was taken from:
+* - LKD 3ed.
+* - RBT kernel API - https://www.infradead.org/~mchehab/kernel_docs/unsorted/rbtree.html
+* - Character driver example - https://gist.github.com/ksvbka/0cd1a31143c1003ce6c7
+* - Max's Holub hivemod.c - https://bit.ly/2o3rmvl
+*/
+
+
+#include <linux/module.h>	// required by all modules
+#include <linux/kernel.h>	// required for sysinfo
+#include <linux/init.h>		// used by module_init, module_exit macros
+#include <linux/jiffies.h>	// where jiffies and its helpers reside
+#include <linux/fs.h>
+#include <linux/rbtree.h>
+#include <linux/errno.h>
+#include <linux/slab.h>
+#include <linux/cdev.h>
+#include <linux/ioctl.h>	// used by ioctl calls
+#include <linux/cdev.h>		// used by char dev
+#include <linux/uaccess.h>	// used by copy_to/from_user 
+
+MODULE_DESCRIPTION("Character device demo");
+MODULE_AUTHOR("rtchoke");
+MODULE_VERSION("0.1");
+MODULE_LICENSE("Dual MIT/GPL");
+
+#define UPD_BUFFER _IOW('D', 'i', unsigned long *)
+#define ADD_MGC_PHR _IOW('C', 'k', unsigned long *)
```

Comment:
Не понятно, почему выбраны именно такие `type`. Как указано [тут](https://www.circlemud.org/jelson/software/fusd/docs/node31.html):

> ``type`` should be chosen so as not to conflict with other drivers that might be _listening_ to the same file descriptor. (Inside the kernel, for example, the TCP and IP stacks use distinct numbers since an ioctl sent to a socket file descriptor might be examined by both stacks.)